src/net_processing.cpp

```diff
@@ -1944,6 +1944,25 @@ static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_pa
     return true;
 }
 
+/**
+ * Do a lookup on the block filter index. The lookup may fail erroneously if the filter index, which
+ * is updated asynchronously, has not been synchronized with the net processing thread. In that
+ * case, block for a short time until the filter index is updated, then retry the lookup.
+ */
+static bool QueryFilterIndexWithRetry(BaseIndex* index, bool& in_sync, std::function<bool()> fn)
```

Comment:
But you may have race condition between client and server where block isn't fully processed before request is handled, so client would have to bear this case anyway

Reply:
The node is only guaranteed to serve a block filter/filter header if it has already advertised the block through a `headers` or `block` message to the requesting peer.